### PR TITLE
lsquic

### PR DESCRIPTION
**Pull request by turuslan**
- integrate `lsquic` to debug problems with `nexus` #227

### CODE REVIEW

File: src/transport/quic/listener.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  bool QuicListener::isClosed() const {
+    return not server_;
+  }
+
+  outcome::result<void> QuicListener::close() {
+    server_.reset();
+    return outcome::success();
+  }
+}  // namespace libp2p::transport
```
Comment (kamilsa) [inline src/transport/quic/listener.cpp:61]: Add empty line

File: src/transport/quic/stream.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  outcome::result<Multiaddress> QuicStream::remoteMultiaddr() const {
+    return conn_->remoteMultiaddr();
+  }
+
+  void QuicStream::onClose() {
+    stream_ctx_ = nullptr;
+  }
+}  // namespace libp2p::connection
```
Comment (kamilsa) [inline src/transport/quic/stream.cpp:128]: empty line

File: src/transport/quic/transport.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    return std::make_shared<QuicListener>(
+        io_context_, ssl_context_, mux_config_, key_codec_, std::move(handler));
+  }
+
+  bool QuicTransport::canDial(const Multiaddress &ma) const {
+    return detail::asQuic(ma).has_value();
+  }
+
+  peer::ProtocolName QuicTransport::getProtocolId() const {
+    return "/quic/1.0.0";
```
Comment (kamilsa) [inline src/transport/quic/transport.cpp:80]: I see that lite-p2p is probably using different string. Can you please ensure we are compatible?

https://github.com/paritytech/litep2p/blob/master/README.md?plain=1#L61

File: src/transport/quic/engine.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        self->on_accept_(res.value());
+      }
+    };
+    stream_if.on_new_stream = +[](void *void_self, lsquic_stream_t *stream) {
+      auto self = static_cast<Engine *>(void_self);
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      auto conn_ctx = reinterpret_cast<ConnCtx *>(
+          lsquic_conn_get_ctx(lsquic_stream_conn(stream)));
+      // NOLINTNEXTLINE(cppcoreguidelines-owning-memory)
+      auto stream_ctx = new StreamCtx{self, stream};
```
Comment (kamilsa) [inline src/transport/quic/engine.cpp:119]: Is it safe to use new here instead of smart pointers?

**Review by kamilsa — APPROVED** [review]

File: src/transport/quic/transport.cpp (hunk 3 above)
Comment (turuslan) [reply src/transport/quic/transport.cpp:80]: These are not related.
`getProtocolId()` is debug string which isn't used/sent anywhere.
Referenced `listen_addresses` is multiaddress.

File: src/transport/quic/transport.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#include <libp2p/security/tls/ssl_context.hpp>
```
Comment (igor-egorov) [inline src/transport/quic/transport.cpp:1]:
```suggestion
/**
 * Copyright Quadrivium LLC
 * All Rights Reserved
 * SPDX-License-Identifier: Apache-2.0
 */

#include <libp2p/security/tls/ssl_context.hpp>
```

File: test/libp2p/transport/quic_test.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  void inject(T &t) {
+    t = injector.create<T>();
+  }
+
+  Injector injector;
+  std::shared_ptr<Host> host;
+  std::shared_ptr<Stream> stream;
+};
+
+TEST(Quic, Test) {
```
Comment (igor-egorov) [inline test/libp2p/transport/quic_test.cpp:47]: Test description is missing

**Review by igor-egorov — APPROVED** [review]

File: src/transport/quic/connection.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  outcome::result<Multiaddress> QuicConnection::remoteMultiaddr() {
+    return remote_;
+  }
+
+  outcome::result<Multiaddress> QuicConnection::localMultiaddr() {
+    return local_;
+  }
+
+  outcome::result<PeerId> QuicConnection::localPeer() const {
+    throw std::logic_error{"QuicConnection::localPeer deprecated"};
```
[the remark this answers is not in the document]
Comment (turuslan) [reply src/transport/quic/connection.cpp:77]: This method is not used.
`localPeer() == Host.getId()` (if program has one `Host`).

File: include/libp2p/transport/quic/engine.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  struct Engine;
+  struct ConnCtx;
+  struct StreamCtx;
+
+  using OnConnect =
+      std::function<void(outcome::result<std::shared_ptr<QuicConnection>>)>;
+  struct Connecting {
+    boost::asio::ip::udp::endpoint remote;
+    PeerId peer;
```
[the remark this answers is not in the document]
Comment (turuslan) [reply include/libp2p/transport/quic/engine.hpp:47]: Should we rename to `endpoint` or spell `remote_endpoint`?

File: src/transport/quic/engine.cpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+      : io_context_{std::move(io_context)},
+        ssl_context_{std::move(ssl_context)},
+        key_codec_{std::move(key_codec)},
+        socket_{std::move(socket)},
+        timer_{*io_context_},
+        socket_local_{socket_.local_endpoint()},
+        local_{detail::makeQuicAddr(socket_local_).value()} {
+    socket_.non_blocking(true);
+
+    lsquicInit();
```
[the remark this answers is not in the document]
Comment (turuslan) [reply src/transport/quic/engine.cpp:31]: `lsquicInit()` fails if boringssl runs out of resources, which would affect many unrelated components.
3 instances of `Engine` (server, ip4 client, ip6 client) is constructed during node initialization.

File: include/libp2p/transport/quic/listener.hpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    bool canListen(const Multiaddress &ma) const override;
+    outcome::result<Multiaddress> getListenMultiaddr() const override;
+
+   private:
+    std::shared_ptr<boost::asio::io_context> io_context_;
+    std::shared_ptr<boost::asio::ssl::context> ssl_context_;
+    muxer::MuxedConnectionConfig mux_config_;
+    std::shared_ptr<crypto::marshaller::KeyMarshaller> key_codec_;
+    TransportListener::HandlerFunc handler_;
+    std::shared_ptr<lsquic::Engine> server_;
```
[the remark this answers is not in the document]
Comment (turuslan) [reply include/libp2p/transport/quic/listener.hpp:47]: Should we rename to `engine_` or spell `server_engine_`?

**Pull request merged**